src/lex/mod.rs

```diff
@@ -353,14 +348,18 @@ impl Lexer {
         if hex {
             if self.match_next(b'p') || self.match_next(b'P') {
                 if !is_digit(self.peek()) {
-                    return Err(String::from("exponent for floating literal has no digits"));
+                    // I'm leaving this as a generic error because I will get
+                    // rid of it with hexponent.
+                    return Err(LexError::Generic(
+                        "exponent for floating literal has no digits".to_string(),
+                    ));
                 }
                 buf.push('p');
                 buf.push(self.next_char().unwrap() as char);
             }
         } else if self.match_next(b'e') || self.match_next(b'E') {
             if !is_digit(self.peek()) {
-                return Err(String::from("exponent for floating literal has no digits"));
+                return Err(LexError::FloatExponentMissingDigits);
```

Comment:
I think this can just be `ExponentMissingDigits`, having an exponent automatically makes it a float.